File: Swift/ransom-note.swift
class Solution {
    func canConstruct1(_ ransomNote: String, _ magazine: String) -> Bool {
        guard magazine.count >= ransomNote.count else {
            return false
        }
        var dict = [Character: Int]()

        for char in magazine {
            if let count = dict[char] {
                dict[char] = count + 1
            } else {
                dict[char] = 1
            }
        }

        for char in ransomNote {
            if let count = dict[char], count > 0 {
                dict[char] = count - 1
            } else {
                return false
            }
        }

        return true
    }

    func canConstruct2(_ ransomNote: String, _ magazine: String) -> Bool {
        guard magazine.count >= ransomNote.count else {
            return false
        }
        var magazine = magazine

        for char in ransomNote {
            if let index = magazine.index(of: char) {
                magazine.remove(at: index)
            } else {
                return false
            }
        }

        return true
    }
}

File: Python/ransom-note.py
class Solution:
    def canConstruct1(self, ransomNote, magazine):
        """
        :type ransomNote: str
        :type magazine: str
        :rtype: bool
        """
        if len(ransomNote) > len(magazine):
            return False

        d = dict()

        for char in magazine:
            if char in d:
                d[char] = d[char] + 1
            else:
                d[char] = 1

        for char in ransomNote:
            if char in d and d[char] > 0:
                d[char] = d[char] - 1
            else:
                return False

        return True

    def canConstruct2(self, ransomNote, magazine):
        """
        :type ransomNote: str
        :type magazine: str
        :rtype: bool
        """
        if len(ransomNote) > len(magazine):
            return False

        for char in ransomNote:
            if char in magazine:
                magazine = magazine.replace(char, '', 1)
            else:
                return False

        return True
